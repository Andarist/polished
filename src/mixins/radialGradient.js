// @flow
import messageHandlers, {
  customRules,
  typeChecks,
} from '../internalHelpers/_messageHandlers'

/** */
type RadialGradientConfiguration = {
  colorStops: Array<string>,
  extent?: string,
  fallback?: string,
  position?: string,
  shape?: string,
}

function parseFallback(colorStops: Array<string>) {
  return colorStops[0].split(' ')[0]
}

function constructGradientValue(
  literals: Array<string>,
  ...substitutions: Array<string>
) {
  let template = ''
  for (let i = 0; i < literals.length; i += 1) {
    template += literals[i]
    // Adds leading coma if properties preceed color-stops
    if (
      i === 3 &&
      substitutions[i] &&
      (substitutions[0] || substitutions[1] || substitutions[2])
    ) {
      template = template.slice(0, -1)
      template += `, ${substitutions[i]}`
      // No trailing space if color-stops is the only param provided
    } else if (
      i === 3 &&
      substitutions[i] &&
      (!substitutions[0] && !substitutions[1] && !substitutions[2])
    ) {
      template += `${substitutions[i]}`
      // Only adds substitution if it is defined
    } else if (substitutions[i]) {
      template += `${substitutions[i]} `
    }
  }
  return template.trim()
}

/**
 * CSS for declaring a radial gradient, including a fallback background-color. The fallback is either the first color-stop or an explicitly passed fallback color.
 *
 * @example
 * // Styles as object usage
 * const styles = {
 *   ...radialGradient({
 *     colorStops: ['#00FFFF 0%', 'rgba(0, 0, 255, 0) 50%', '#0000FF 95%'],
 *     extent: 'farthest-corner at 45px 45px',
 *     position: 'center',
 *     shape: 'ellipse',
 *   })
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   ${radialGradient({
 *     colorStops: ['#00FFFF 0%', 'rgba(0, 0, 255, 0) 50%', '#0000FF 95%'],
 *     extent: 'farthest-corner at 45px 45px',
 *     position: 'center',
 *     shape: 'ellipse',
 *   })}
 *`
 *
 * // CSS as JS Output
 *
 * div: {
 *   'backgroundColor': '#00FFFF',
 *   'backgroundImage': 'radial-gradient(center ellipse farthest-corner at 45px 45px, #00FFFF 0%, rgba(0, 0, 255, 0) 50%, #0000FF 95%)',
 * }
 */

function radialGradient(config: RadialGradientConfiguration) {
  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'production') {
    if (
      messageHandlers('mixins/radialGradient.js', {
        // eslint-disable-next-line prefer-rest-params
        arrityCheck: { args: arguments, exactly: 1, throw: true },
        typeChecks: {
          param: config,
          type: 'object',
          throw: true,
          required: 'requires a config object as its only parameter. However, you did not provide one.',
        },
      })
    ) {
      return {}
    }
  }

  const { colorStops, extent, fallback, position, shape } = config

  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'production') {
    if (
      !typeChecks('mixins/radialGradient.js', [
        {
          param: colorStops,
          type: 'array',
          required: 'expects an array of at least 2 color-stops.',
          throw: true,
        },
        { param: extent, type: 'string' },
        { param: fallback, type: 'string' },
        { param: position, type: 'string' },
        { param: shape, type: 'string' },
      ]) ||
      !customRules('mixins/radialGradient.js', {
        enforce: colorStops.length > 1,
        msg: `expects an array of at least 2 color-stops. However, the one you provided only had ${colorStops.length}.`,
      })
    ) {
      return {}
    }
  }

  return {
    backgroundColor: fallback || parseFallback(colorStops),
    backgroundImage: constructGradientValue`radial-gradient(${position}${shape}${extent}${colorStops.join(', ')})`,
  }
}

export default radialGradient
