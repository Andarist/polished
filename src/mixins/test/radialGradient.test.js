// @flow
import radialGradient from '../radialGradient'

describe('radialGradient', () => {
  beforeAll(() => {
    global.console = {
      error: jest.fn(),
      warn: jest.fn(),
    }
  })

  afterEach(() => {
    // eslint-disable-next-line no-console
    console.error.mockClear()
    // eslint-disable-next-line no-console
    console.warn.mockClear()
  })

  it('returns the correct object when only passed two color stops, including parsed fallback with no percentage', () => {
    expect({
      ...radialGradient({
        colorStops: ['#fff', '#000'],
      }),
    }).toMatchSnapshot()
  })

  it('returns the correct object when passed extent, shape, and position, including parsed fallback with percentage', () => {
    expect({
      ...radialGradient({
        colorStops: ['#00FFFF 0%', 'rgba(0, 0, 255, 0) 50%', '#0000FF 95%'],
        extent: 'farthest-corner at 45px 45px',
        position: 'center',
        shape: 'ellipse',
      }),
    }).toMatchSnapshot()
  })

  it('returns the correct object when passed extent and shape', () => {
    expect({
      ...radialGradient({
        colorStops: ['#00FFFF 0%', 'rgba(0, 0, 255, 0) 50%', '#0000FF 95%'],
        extent: 'farthest-corner at 45px 45px',
        shape: 'ellipse',
      }),
    }).toMatchSnapshot()
  })

  it('returns the correct object when passed just extent', () => {
    expect({
      ...radialGradient({
        colorStops: ['#00FFFF 0%', 'rgba(0, 0, 255, 0) 50%', '#0000FF 95%'],
        extent: 'farthest-corner at 45px 45px',
      }),
    }).toMatchSnapshot()
  })

  it('properly overrides the fallback when it is passed', () => {
    expect({
      ...radialGradient({
        colorStops: ['#00FFFF 0%', 'rgba(0, 0, 255, 0) 50%', '#0000FF 95%'],
        extent: 'farthest-corner at 45px 45px',
        fallback: '#FFF',
      }),
    }).toMatchSnapshot()
  })

  it('should throw a warning when passed more than one parameter', () => {
    radialGradient(
      {
        colorStops: ['#00FFFF 0%', 'rgba(0, 0, 255, 0) 50%', '#0000FF 95%'],
        extent: 'farthest-corner at 45px 45px',
        fallback: '#FFF',
      },
      1,
    )
    // eslint-disable-next-line no-console
    expect(console.warn.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when not provided any parameters', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    radialGradient()
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed a non-object value', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    radialGradient(1)
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when not passed a value for colorStops', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    radialGradient({})
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed a non-array value for colorStops', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    radialGradient({ colorStops: true })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed an array with less than two values for colorStops', () => {
    radialGradient({ colorStops: ['#00FFFF 0%'] })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed a non-string value for extent', () => {
    radialGradient({
      colorStops: ['#00FFFF 0%', 'rgba(0, 0, 255, 0) 50%', '#0000FF 95%'],
      // $FlowIgnoreNextLine since the coming is invalid code, flow complains
      extent: true,
      position: 'center',
      shape: 'ellipse',
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed a non-string value for position', () => {
    radialGradient({
      colorStops: ['#00FFFF 0%', 'rgba(0, 0, 255, 0) 50%', '#0000FF 95%'],
      // $FlowIgnoreNextLine since the coming is invalid code, flow complains
      position: true,
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed a non-string value for shape', () => {
    radialGradient({
      colorStops: ['#00FFFF 0%', 'rgba(0, 0, 255, 0) 50%', '#0000FF 95%'],
      // $FlowIgnoreNextLine since the coming is invalid code, flow complains
      shape: true,
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })
})
