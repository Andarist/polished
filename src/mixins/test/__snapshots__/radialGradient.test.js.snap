exports[`radialGradient properly overrides the fallback when it is passed 1`] = `
Object {
  "backgroundColor": "#FFF",
  "backgroundImage": "radial-gradient(farthest-corner at 45px 45px, #00FFFF 0%, rgba(0, 0, 255, 0) 50%, #0000FF 95%)",
}
`;

exports[`radialGradient returns the correct object when only passed two color stops, including parsed fallback with no percentage 1`] = `
Object {
  "backgroundColor": "#fff",
  "backgroundImage": "radial-gradient(#fff, #000)",
}
`;

exports[`radialGradient returns the correct object when passed extent and shape 1`] = `
Object {
  "backgroundColor": "#00FFFF",
  "backgroundImage": "radial-gradient(ellipse farthest-corner at 45px 45px, #00FFFF 0%, rgba(0, 0, 255, 0) 50%, #0000FF 95%)",
}
`;

exports[`radialGradient returns the correct object when passed extent, shape, and position, including parsed fallback with percentage 1`] = `
Object {
  "backgroundColor": "#00FFFF",
  "backgroundImage": "radial-gradient(center ellipse farthest-corner at 45px 45px, #00FFFF 0%, rgba(0, 0, 255, 0) 50%, #0000FF 95%)",
}
`;

exports[`radialGradient returns the correct object when passed just extent 1`] = `
Object {
  "backgroundColor": "#00FFFF",
  "backgroundImage": "radial-gradient(farthest-corner at 45px 45px, #00FFFF 0%, rgba(0, 0, 255, 0) 50%, #0000FF 95%)",
}
`;

exports[`radialGradient should throw an error when not passed a value for colorStops 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/radialGradient.js --%c

radialGradient %crequires a %c1st parameter%c(%ccolorStops%c: %carray%c). However, you did not pass %ccolorStops%c.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#radialGradient%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`radialGradient should throw an error when not provided any parameters 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/radialGradient.js --%c

radialGradient %crequires a %c1st parameter%c(%ccolorStops%c: %carray%c). However, you did not pass %ccolorStops%c.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#radialGradient%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`radialGradient should throw an error when passed a non-array value for colorStops 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/radialGradient.js --%c

radialGradient %cexpects a %c1st parameter%c(%ccolorStops%c: %carray%c). However, you passed (%c\'true\'%c:%cboolean%c) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#radialGradient%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`radialGradient should throw an error when passed a non-string value for extent 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/radialGradient.js --%c

radialGradient %cexpects a %c2nd parameter%c(%cextent%c: %cstring%c). However, you passed (%c\'true\'%c:%cboolean%c) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#radialGradient%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`radialGradient should throw an error when passed a non-string value for fallback 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/radialGradient.js --%c

radialGradient %cexpects a %c3rd parameter%c(%cfallback%c: %cstring%c). However, you passed (%c\'true\'%c:%cboolean%c) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#radialGradient%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`radialGradient should throw an error when passed a non-string value for position 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/radialGradient.js --%c

radialGradient %cexpects a %c4th parameter%c(%cposition%c: %cstring%c). However, you passed (%c\'true\'%c:%cboolean%c) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#radialGradient%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`radialGradient should throw an error when passed a non-string value for shape 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/radialGradient.js --%c

radialGradient %cexpects a %c5th parameter%c(%cshape%c: %cstring%c). However, you passed (%c\'true\'%c:%cboolean%c) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#radialGradient%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`radialGradient should throw an error when passed an array with less than two values for colorStops 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/radialGradient.js --%c

radialGradient %cexpects a %c1st parameter%c(%ccolorStops%c: %carray%c). However, you passed (%c\'#00FFFF 0%\'%c:%cobject%c) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#radialGradient%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;
